cdc/log.cc

```diff
@@ -538,23 +563,136 @@ class transformer final {
                                 if (view.is_live_and_has_ttl()) {
                                     ttl = view.ttl();
                                 }
-                            } else {
-                                op = column_op::del;
-                            }
-
-                            values[0] = data_type_for<column_op_native_type>()->decompose(data_value(static_cast<column_op_native_type>(op)));
-                            res.set_cell(log_ck, *dst, atomic_cell::make_live(*dst->type, ts, tuple_type_impl::build_value(values)));
-
-                            if (pirow && pirow->has(cdef.name_as_text())) {
-                                values[0] = data_type_for<column_op_native_type>()->decompose(data_value(static_cast<column_op_native_type>(column_op::set)));
-                                values[1] = pirow->get_blob(cdef.name_as_text());
-
-                                assert(std::addressof(res.partition().clustered_row(*_log_schema, *pikey)) != std::addressof(res.partition().clustered_row(*_log_schema, log_ck)));
-                                assert(pikey->explode() != log_ck.explode());
-                                res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, tuple_type_impl::build_value(values)));
-                            }
+                            } 
                         } else {
-                            cdc_log.warn("Non-atomic cell ignored {}.{}:{}", _schema->ks_name(), _schema->cf_name(), cdef.name_as_text());
+                            auto mv = cell.as_collection_mutation();
+                            auto now = gc_clock::time_point::min();
+                            std::optional<column_op> oop;
+                            op = column_op::add;
+
+                            std::vector<bytes> buf;
+                            values[1] = mv.with_deserialized(*cdef.type, [&](collection_mutation_view_description view) -> bytes_opt {
+                                if (view.tomb) {
+                                    // there is a tombstone with timestamp before this mutation.
+                                    // this is how a assign collection = <value> is represented.
+                                    oop = column_op::set;
+                                }
+                                auto process = [&](auto value_callback) {
+                                    for (auto& [key, value] : view.cells) {
+                                        auto live = value.is_live(view.tomb, now, false);
```

Comment:
We cannot refer to the current clock time there, this will give non-deterministic results, and we lose replayability!
If someone writes a live cell to the base table with ttl = 1 say, and the cell expires by the time we process it here in CDC code, we'll show it in the log as a dead cell (null) --- instead it should be a live cell (with a non-null value) with ttl 1.

I guess you want to distinguish here between:
1. a dead cell, i.e. assignment to null; this happens when `!value.is_live()`
2. an alive cell, perhaps with ttl or without, doesn't matter; this happens when `value.is_live()`

Reply:
Yes and no. The problem/reason why the tombstone is included is how a collection assignment is done: A tombstone is written for the column (the "tomb" here), and then we add cells. However, technically, it would be "legal" for any of these cells to be timestamped/tombstoned so that they fall _before_ the column tomb. It is obviously a ridiculous case, but we are only operating on raw mutations here... 
I agree with the reasoning that we might race the clock for an actual user provided ttl etc, and perhaps the esoteric case should just be ignored...

Reply:
If cells are added with timestamp X, the tombstone is added with timestamp X-1. E.g. cql3/maps.cc, `maps::setter::execute`:
```
        mut.tomb = params.make_tombstone_just_before();
```
where `update_parameters::make_tombstone_just_before`:
```
    tombstone make_tombstone_just_before() const {
        return {_timestamp - 1, _local_deletion_time};
    }
```
(`_local_deletion_time` is the `expiry` value, we ignore it)

So if these cells and this tombstone were made in a single `column = value` assignment, the tombstone will never kill the cells.

If `tombstone's timestamp` won't be `cells' timestamp - 1`, I'll split them to separate mutations (I'm working on that splitting stuff right now). Don't worry about that.

Hell, maybe I'll even split the tombstone to a separate mutation, so a `column = value` statement for collection columns will always give two mutations: one of them does `column = null` with timestamp `X - 1`, and the other does a bunch of `column[idx] = val` with timestamp `X`. I have to decide what to do exactly there yet.

Reply:
Yes, assuming the mutation is well-formed. But otoh, we should probably just assume that, so I agree with your statement.